keras_cv/layers/preprocessing3d/frustum_random_point_feature_noise.py

```diff
@@ -0,0 +1,121 @@
+
+from keras_cv.layers.preprocessing3d import base_augmentation_layer_3d
+from keras_cv.ops import point_cloud
+
+POINT_CLOUDS = base_augmentation_layer_3d.POINT_CLOUDS
+BOUNDING_BOXES = base_augmentation_layer_3d.BOUNDING_BOXES
+POINTCLOUD_LABEL_INDEX = base_augmentation_layer_3d.POINTCLOUD_LABEL_INDEX
+POINTCLOUD_FEATURE_INDEX = base_augmentation_layer_3d.POINTCLOUD_FEATURE_INDEX
+
+
+class FrustumRandomPointFeatureNoise(
+    base_augmentation_layer_3d.BaseAugmentationLayer3D
+):
+    """A preprocessing layer which randomly add noise to point features within a randomly generated frustum during training.
+
+    This layer will randomly select a point from the point cloud as the center of a frustum then generate a frustum based
+    on r_distance, theta_width, and phi_width. Uniformly sampled features noise from [1-max_noise_level, 1+max_noise_level] will be multiplied
+    to points inside the selected frustum. Here, we perturbe point features other than (x, y, z, class).
+    The point_clouds tensor shape must be specific and cannot be dynamic.
+    During inference time, the output will be identical to input. Call the layer with `training=True` to add noise to the input points.
+
+    Input shape:
+      point_clouds: 3D (multi frames) float32 Tensor with shape
+        [num of frames, num of points, num of point features].
+        The first 5 features are [x, y, z, class, range].
+      bounding_boxes: 3D (multi frames) float32 Tensor with shape
+        [num of frames, num of boxes, num of box features].
+        The first 7 features are [x, y, z, dx, dy, dz, phi].
+
+    Output shape:
+      A dictionary of Tensors with the same shape as input Tensors.
+
+    Arguments:
+      r_distance: A float scalar sets the starting distance of a frustum.
+      theta_width: A float scalar sets the theta width of a frustum.
+      phi_width: A float scalar sets the phi width of a frustum.
+      max_noise_level: A float scalar sets the probability threshold for dropping the points.
+
+    """
+
+    def __init__(
+        self, r_distance, theta_width, phi_width, max_noise_level=None, **kwargs
+    ):
+        super().__init__(**kwargs)
+
+        if r_distance < 0:
+            raise ValueError("r_distance must be >=0.")
+        if theta_width < 0:
+            raise ValueError("theta_width must be >=0.")
+        if phi_width < 0:
+            raise ValueError("phi_width must be >=0.")
+        max_noise_level = max_noise_level if max_noise_level else 0.0
+        if max_noise_level < 0 or max_noise_level > 1:
+            raise ValueError("max_noise_level must be >=0 and <=1.")
+
+        self._r_distance = r_distance
+        self._theta_width = theta_width
+        self._phi_width = phi_width
+        self._max_noise_level = max_noise_level
+
+    def get_random_transformation(self, point_clouds, **kwargs):
+        # Randomly select a point from the first frame as the center of the frustum.
+        num_valid_points = tf.cast(
```

Comment:
This looks like a lot of ~duplicate code from the FrustrumRandomDropping layer. 

Maybe we should write a util to build a frustrum and identify which points are in it?

Reply:
Good point! We could clean it up in a future PR. The redundant codes are random sample a frustum center. 
Most of the frustum related utils are in this function `point_cloud.within_a_frustum`